lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,253 @@
+"""
+Models used for robust grading.
+
+Robust grading allows student scores to be saved per-subsection independent
+of any changes that may occur to the course after the score is achieved.
+"""
+
+from django.db import models
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from opaque_keys.edx.locator import BlockUsageLocator
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+from hashlib import sha256
+import json
+
+
+class VisibleBlocks(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a given subsection at the time they are
+    used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored in the blocks_json field. A
+    hash of this json array is used for lookup purposes.
+    """
+    _blocks_json = models.TextField(db_column="blocks_json")
+    hashed = models.CharField(max_length=32, primary_key=True)
+
+    def __unicode__(self):
+        """
+        String representation of this model.
+        """
+        return self.hashed
+
+    @classmethod
+    def create(cls, blocks):
+        """
+        Creates a new VisibleBlocks model. Argument 'blocks' should be an iterable collection of BlockRecords.
+        """
+        # Start by sorting the blocks to ensure equality regardless of how the blocks are ordered
+        sorted_blocks = sorted(
+            blocks,
+            key=lambda block: '{}{}{}'.format(block.block_key, block.max_score, block.weight)
```

Comment:
I would vote for keeping the sorting - in favor of robustness.

It seems fragile to assume that the serialized order of the blocks is the desired order of viewing the blocks in the UI.  A topological sort would provide a deterministic order for DAGs - but not necessarily how we want to display them in the UI.  In the UI, we've been presenting them hierarchically by subsections.